android-core/src/main/java/com/mparticle/MParticle.java

```diff
@@ -261,6 +281,38 @@ public static void setInstance(@Nullable MParticle instance) {
         MParticle.instance = instance;
     }
 
+    /**
+     * Switch the SDK to a new API key and secret.
+     * Will first batch all events that have not been sent to mParticle into upload records,
+     * then all SDK state including user defaults, database (except uploads), etc will be
+     * completely reset. After that, {@link #start(MParticleOptions)} )} will be called with
+     * the new key and secret and the SDK will initialize again as if it is a new app launch.
+     *
+     @param options Required to initialize the SDK properly
+     */
+    public static void switchWorkspace(@NonNull MParticleOptions options) {
+        if (instance != null) {
+            // End session if active
+            if (instance.isSessionActive()) {
+                instance.endSession();
+            }
+
+            // Batch any remaining messages into upload records
+            try {
+                instance.mMessageManager.mUploadHandler.prepareMessageUploads(instance.mConfigManager.getUploadSettings());
```

Comment:
key factor in making threading work in the android SDK (and ios sdk for that matter..) is to make sure things run sequentially.

as you have it now - this will happen:

1a. customer calls `mparticle.logEvent("foo")`
1b async invoke message handler thread to store "foo"
2. then `mparticle.switchWorkpace("bar")`

and you'll prepare message uploads in (2) before (1b) ever gets around to even storing the "foo" event in the message table. actually you'll have this problem with the call to "endSession" above too since you're not waiting for that to store the resulting message.

so - you need to wait on the messages thread to make sure every even to this point has been stored before we prepare and delete etc.

but wait..there's more. the actual call to prepare should probably happen on the uploads thread, otherwise you'll end up with this method running simultanously with natural/normal uploads and a mess of inconsistent/unpredictable state.

there's some precedence throughout the code base of doing things like this:

- MessageManager#doUpload() - which forces a message into the message handler and then into the upload handler just to make sure we do a full pass through both.
- IdentityApi which continuously uses `mBackgroundHandler` (which is actually the upload handler!) to ensure that identity requests are synchronous to uploads
- `KitManagerImpl#runOnKitThread` for another example of just using another thread

there's a few ways to code this but ultimately I think you need to accomplish:
- ensuring that everything currently queued on the message thread completes before you prepare this 
- ensure that you don't concurrently modify the database/state/etc with both the message thread and upload thread

Reply:
Please see my latest commit, I believe it should address these concerns